### PR TITLE
executor: store column's fieldType in executor

### PR DESCRIPTION
**Pull request by XuHuaiyu**
As the title says.
PTAL @zz-jason  @coocood

### CODE REVIEW

File: expression/schema.go [hunk 1]
```diff
@@ -210,11 +211,13 @@ func (s *Schema) ColumnsByIndices(offsets []int) []*Column {
 
 // GetTypes creates a field type slice.
 func (s *Schema) GetTypes() []*types.FieldType {
-	fields := make([]*types.FieldType, len(s.Columns))
-	for i := range fields {
-		fields[i] = s.Columns[i].RetType
+	if s.fieldTypes == nil {
```
Comment (winkyao) [inline expression/schema.go]: is GetTypes thread-safe?

Comment (zz-jason) [reply expression/schema.go]: nice catch, it's not thread safe but can be called concurrently.

Comment (winkyao) [reply expression/schema.go]: init the buf in the NewSchema is ok.

**Comment by winoros** [comment]
Please modify `Schema.Append` method.

**Comment by ngaut** [comment]
Please post bench results.

**Comment by shenli** [comment]
Please fix CI.

**Comment by XuHuaiyu** [comment]
on my machine:
bench result:
before: 
BenchmarkJoin-4             1000           1459172 ns/op          657269 B/op       9255 allocs/op

after:
BenchmarkJoin-4             1000           1310514 ns/op          656533 B/op       9241 allocs/op

**Comment by XuHuaiyu** [comment]
PTAL @zz-jason @winkyao

File: expression/schema.go [hunk 2]
```diff
@@ -208,15 +207,6 @@ func (s *Schema) ColumnsByIndices(offsets []int) []*Column {
 	return cols
 }
 
-// GetTypes creates a field type slice.
-func (s *Schema) GetTypes() []*types.FieldType {
```
Comment (lamxTyler) [inline expression/schema.go]: Why remove this? There are some places in your code that can be replaced by this function.

File: executor/executor.go [hunk 3]
```diff
@@ -174,6 +188,7 @@ type Executor interface {
 	Close() error
 	Open(goctx.Context) error
 	Schema() *expression.Schema
+	RetTypes() []*types.FieldType
```
Comment (zz-jason) [inline executor/executor.go]: no need to expose it

File: expression/schema.go (hunk 2 above)
Comment (XuHuaiyu) [reply expression/schema.go]: PTAL @lamxTyler

**Comment by XuHuaiyu** [comment]
ptal @zz-jason

File: executor/builder.go [hunk 4]
```diff
@@ -1081,18 +1081,19 @@ func (b *executorBuilder) buildIndexLookUpJoin(v *plan.PhysicalIndexJoin) Execut
 }
 
 func (b *executorBuilder) buildNewIndexLookUpJoin(v *plan.PhysicalIndexJoin, outerExec Executor) Executor {
+	innerPlan := v.Children()[1-v.OuterIndex]
+	outerTypes, innerTypes := outerExec.retTypes(), innerPlan.Schema().GetTypes()
 	outerFilter, innerFilter := v.LeftConditions, v.RightConditions
-	leftTypes, rightTypes := v.Children()[0].Schema().GetTypes(), v.Children()[1].Schema().GetTypes()
-	outerTypes, innerTypes := leftTypes, rightTypes
+	leftTypes, rightTypes := outerTypes, innerTypes
```
Comment (lamxTyler) [inline executor/builder.go]: Why use new variables? It is not used in the following code anyway.

**Pull request merged**